src/Visualization/Python/ApplyPointwise.py

```diff
@@ -0,0 +1,327 @@
+                f"'{self.output_type}'.")
+
+
+def apply_pointwise(volfiles: Union[spectre_h5.H5Vol,
+                                    Iterable[spectre_h5.H5Vol]],
+                    kernels: Sequence[Kernel]):
+    """Apply pointwise functions to data in the 'volfiles'
+
+    Arguments:
+      volfiles: Iterable of open H5 volume files, or a single H5 volume file.
+        Must be opened in writable mode, e.g. in mode "a". The transformed
+        data will be written back into these files. All observations in these
+        files will be transformed.
+      kernels: List of pointwise transformations to apply to the volume data
+        in the form of 'Kernel' objects.
+    """
+    # Collect all tensor components that we need to apply the kernels
+    all_tensors = {}
+    for kernel in kernels:
+        for tensor_name, tensor_type in zip(kernel.input_names,
+                                            kernel.input_types):
+            if tensor_name in all_tensors:
+                assert all_tensors[tensor_name][0] == tensor_type, (
+                    "Two tensor arguments with the same name "
+                    f"'{tensor_name}' have different types: "
+                    f"'{all_tensors[tensor_name][0]}' and '{tensor_type}'")
+            else:
+                all_tensors[tensor_name] = (tensor_type,
+                                            get_tensor_component_names(
+                                                tensor_name, tensor_type))
+    logger.debug("Input datasets: " + str(list(all_tensors.keys())))
+    logger.info("Output datasets: " +
+                str([kernel.output_name for kernel in kernels]))
+
+    if isinstance(volfiles, spectre_h5.H5Vol):
+        volfiles = [volfiles]
+    for volfile in volfiles:
+        for obs_id in volfile.list_observation_ids():
+            # Load tensor data for all kernels
+            all_tensor_data = {
+                tensor_name: tensor_type(
+                    np.array([
+                        volfile.get_tensor_component(obs_id,
+                                                     component_name).data
+                        for component_name in component_names
+                    ]))
+                for tensor_name, (tensor_type,
+                                  component_names) in all_tensors.items()
+            }
+
+            # Apply kernels
+            # We currently only apply strictly pointwise kernels that need no
+            # information about the element or mesh. To support derivatives we
+            # can slice the tensor data into elements and call those kernels
+            # that request it element-wise, passing them the mesh and Jacobian
+            # as well.
+            for kernel in kernels:
+                kernel_args = [
+                    all_tensor_data[input_name]
+                    for input_name in kernel.input_names
+                ]
+
+                transformed_tensor = kernel.callable(*kernel_args)
```

Comment:
If I read this right, each kernel must return exactly one tensor. Would it be easy to (maybe in a followup pR) allow returning a tuple/list/dict of tensors (whatever is easiest)?

Reply:
Yea I meant to do that but didn't come up with a nice way of specifying the dataset names for the tensors. Maybe returning a dict is the easiest solution.

Reply:
Yea, I would go for a dict :)